core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java

```diff
@@ -1057,6 +1057,19 @@ void testDyadicCollateOperator() {
             + "Was expecting 3 arguments");
   }
 
+  /** Test case for <a href="https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-6813">
+   * [CALCITE-6813] UNNEST infers incorrect nullability for the result when applied to
+   * an array that contains nullable ROW values</a>. */
+  @Test void testUnnestRow() {
+    sql("with orders(data) as\n"
+        + "  (values (ARRAY[ROW(1, 'Alice'), ROW(2, NULL), ROW(NULL, 'Bob'), NULL]))\n"
+        + "select e.EXPR$0\n"
+        + "from orders, UNNEST(orders.data) as e")
+        .type(actualType -> {
```

Comment:
What if all ROW values are not nullable?

Reply:
I assume you mean: "none of the values in any ROW is null."
That case was covered before. ROWs with null values were not handled correctly.